### ISSUE TITLE
ReferenceError: WebSocket is not defined

### ISSUE THREAD

**Issue by takahser** [ReactiveX/rxjs#3942]
## Bug Report

**Current Behavior**
I use `WebSocketSubject` and `webSocket` in my angular universal 6.x project, which works fine on the browser platform. However, when building the SSR stuff, it throws me an error:
`ReferenceError: WebSocket is not defined`.

**Reproduction**
- REPL or Repo link:
https://github.com/takahser/angular-universal-ssr-with-rxjs-websocket

Example code:
```
import { WebSocketSubject, webSocket } from 'rxjs/webSocket';

const socket: WebSocketSubject<any> = webSocket('wss://echo.websocket.org');
socket.subscribe(msg => doSomething(msg));
```

**Expected behavior**
No error

**Environment**
- Runtime: Angular 6
- RxJS version: 6.2.1, 6.2.2 (tested both)
- Loader, build configuration:

```
npm install
npm run build:client-and-server-bundles
npm run build:ssr
npm serve:ssr
```

**Comment by trotyl** [comment]
https://github.com/ReactiveX/rxjs/blob/a3323fc1855f2c0bf119d553004dee5e768a6453/src/internal/observable/dom/WebSocketSubject.ts#L94

Likely this should be `... && typeof WebSocket !== 'undefined'` instead.

**Comment by takahser** [comment]
@trotyl Thanks for your suggestion.
I tried to fix it by:

1. forking the rxjs repo => see https://github.com/takahser/rxjs
2. fixing the bug => `if (!config.WebSocketCtor && typeof WebSocket !== 'undefined') {` 
3. replacing the dependency in my package.json => "rxjs": "takahser/rxjs#00d144092d421d00080a7b44c3672af2b4f3733d",
4. running `rm -rf node_modules/ && npm install`

However, when checking `rxjs/src/internal/observable/dom/WebSocketSubject.ts` in my node_modules, it doesn't contain my changes... why?
I decided to manually change the file within node_modules and running the build commands, but it still throws me the error `ReferenceError: WebSocket is not defined` when accessing the running app...

**Comment by trotyl** [comment]
At least you should tell which line of code (on which project) is causing this.

**Comment by takahser** [comment]
@trotyl `WebSocketSubject` is used here: 

https://github.com/takahser/angular-universal-ssr-with-rxjs-websocket/blob/master/src/app/messaging.service.ts#L5

Please note that the error only occurs after

1. building the server-rendered pages (`npm run build:client-and-server-bundles`) 
2. building the express web server (`npm run webpack:server`)
3. starting the express web server (`npm run serve:ssr`)
4. and open the app in the browser

The webserver bundle is generated using this webpack config:
https://github.com/takahser/angular-universal-ssr-with-rxjs-websocket/blob/master/webpack.server.config.js

Please let me know, if you need more information

**Comment by Zaniyar** [comment]
takahser told me to comment here: I have the same problem.

**Comment by takahser** [comment]
@Zaniyar No, but it seems as this is rather a problem with angular universal and webpack than RxJs. When using SSR, Angular Universal creates a bundle for running the angular app on the node server, but there is natively no websocket implementation on Node. Therefore, it has to be added manually as a dependency (npm package). I tried adding it to the js server bundle (`const WebSocket = require('ws');` manually, which resolves the problem. However, when I add it to the typescript code that gets transcompiled into the js bundle later on, it won't work.

**Comment by CoreyCole** [comment]
I'm trying to get WebSocketSubject working on my nestjs backend. I'm getting the same error:
```
ReferenceError: WebSocket is not defined
    at new WebSocketSubject (.../node_modules/rxjs/src/internal/observable/dom/WebSocketSubject.ts:94:36)
```
I know `WebSocketSubject` depends on the dom/browser's native websocket library. I've made sure that "dom" is included in my tsconfig.

**Comment by CoreyCole** [comment]
I saw [your post](https://stackoverflow.com/questions/51532159/referenceerror-websocket-is-not-defined-when-using-rxjs-websocketsubject-and) on stack overflow. Adding this line to the file I was using `WebSocketSubject` in fixed my problem:
```
// tslint:disable-next-line
(global as any).WebSocket = require('ws');
```
But, looks like a hack to me. Anyone know what is the root of this problem?

**Comment by luillyfe** [comment]
Since `webSocketSubject` is a wrapper around `websocket` object from browsers, it makes perfectly sense  not to be able in the server. No one browser APIs is available when using Angular/rxjs in the server.

**Comment by takahser** [comment]
@CoreyCole exactly.
When running JavaScript in the browser, the JavaScript API provides us with various global objects, such as `window`, `document`, or `WebSocket`, which are not available on the server-side node implementation.
When running JS on the server (node), such as it is the case with Angular Prerendering, certain global browser web APIs won't be available. Therefore, we need to use certain npm packages, that provides us with the lacking functionality (such as `ws` in this case). When using JS, this would be sufficient, but TypeScript will complain, because the types on the `WebSocket` object are mising. Finally, we can convert the untyped object to the type `any`, which can be (as the name implies) anything (`(global as any).WebSocket = require('ws');`). This way, we can satisfy the TypeScript compiler. :)

You can find more examples global variables, that are different on the browser and server implementations [here](http://voidcanvas.com/node-vs-browsers/).
